hal/src/common/dmac/channel.rs

```diff
@@ -0,0 +1,357 @@
+}
+
+impl<C: AnyChannel> AsRef<C> for SpecificChannel<C> {
+    #[inline]
+    fn as_ref(&self) -> &C {
+        // SAFETY: This is guaranteed to be safe, because C == SpecificChannel<C>
+        unsafe { mem::transmute(self) }
+    }
+}
+
+impl<C: AnyChannel> AsMut<C> for SpecificChannel<C> {
+    #[inline]
+    fn as_mut(&mut self) -> &mut C {
+        // SAFETY: This is guaranteed to be safe, because C == SpecificChannel<C>
+        unsafe { mem::transmute(self) }
+    }
+}
+
+//==============================================================================
+// Channel
+//==============================================================================
+/// DMA channel, capable of executing
+/// [`Transfer`](super::transfer::Transfer)s.
+pub struct Channel<Id: ChId, S: Status> {
+    _id: PhantomData<Id>,
+    _status: S,
+}
+
+#[inline]
+pub(crate) fn new_chan<Id: ChId>(_id: PhantomData<Id>) -> Channel<Id, Uninitialized> {
+    Channel {
+        _id,
+        _status: Uninitialized,
+    }
+}
+
+/// These methods may be used on any DMA channel in any configuration
+impl<Id: ChId, S: Status> Channel<Id, S> {
+    /// Set channel ID and run the closure. A closure is needed to ensure
+    /// the registers are accessed in an interrupt-safe way, as the SAMD21
+    /// DMAC is a little funky - It requires setting the channel number in
+    /// the CHID register, then access the channel control registers.
+    /// If an interrupt were to change the CHID register, we would be faced
+    /// with undefined behaviour.
+    #[cfg(any(feature = "samd11", feature = "samd21"))]
+    fn with_chid<F: Fn(&mut DMAC)>(&mut self, dmac: &mut DMAC, fun: F) {
+        cortex_m::interrupt::free(|_| {
+            // SAFETY: this is actually safe as long as we write a correct channel number to
+            // the CHID register
+            unsafe {
+                dmac.chid.modify(|_, w| w.id().bits(Id::U8));
+            };
+
+            fun(dmac);
+        });
+    }
+
+    /// Set channel ID and run the closure. A closure is needed to ensure
+    /// the registers are accessed in an interrupt-safe way, as the SAMD21
+    /// DMAC is a little funky. For the SAMD51/SAMEx, we simply take a reference
+    /// to the correct channel number and run the closure on that.
+    #[cfg(feature = "min-samd51g")]
+    #[inline]
+    fn with_chid<F: Fn(&CHANNEL)>(&mut self, dmac: &mut DMAC, fun: F) {
+        let mut ch = &dmac.channel[Id::USIZE];
+        fun(&mut ch);
+    }
+
+    /// Configure the DMA channel so that it is ready to be used by a
+    /// [`Transfer`](super::transfer::Transfer).
+    ///
+    /// # Return
+    ///
+    /// A `Channel` with a `Ready` status
+    pub fn init(
+        mut self,
+        controller: &mut DmaController,
+        lvl: PriorityLevel,
+        enable_interrupts: bool,
+    ) -> Channel<Id, Ready> {
+        // SAFETY: this is safe because we only mutably borrow dmac once.
+        let dmac = unsafe { controller.dmac_mut() };
```

Comment:
Instead of using the `unsafe` function `dmac_mut` everywhere, you might consider making the `dmac` field `pub(super)`. I think that would be cleaner and remove some of the `unsafe` noise.